### ISSUE TITLE
Question: How to map arguments to calls to a repo

### ISSUE THREAD

**Issue by Geraint87** [adamabdelhamed/PowerArgs#62]
I have the class below, with some conditional arguments e.g. a caller can specify a specific Id, or use the `--all` switch to denote all items.

What I really want is to have `--all` map to an end point such as repo.GetItems(), but `-id 3` map to repo.GetItems(3).

So ideally it would be similar to:
```
public class ApplicationArgs
{
	/// stuff

	[ArgShortcut("-i")]
	[ArgRequired(IfNot = "All")]
	[ArgCantBeCombinedWith("All")]
	public IFoo Id { get; set; }

	[ArgShortcut("-a")]
	public IEnumerable<IFoo> All { get; set; }

	// more stuff...
}
```

Is this possible with revivers, or another method? Or how _should_ I structure my application to achieve this same effect with PowerArgs? 

```
public class ApplicationArgs
{
	[HelpHook]
	[ArgPosition(0)]
	[ArgShortcut("-?")]
	public bool Help { get; set; }

	[ArgShortcut("-i")]
	[ArgRequired(IfNot = "All")]
	[ArgCantBeCombinedWith("All")]
	public int Id { get; set; }

	[ArgShortcut("-a")]
	public bool All { get; set; }

	[ArgShortcut("d")]
	[ArgCantBeCombinedWith("History")]
	public DateTime? Date { get; set; }

	[ArgShortcut("-h")]
	[ArgCantBeCombinedWith("All")]
	public bool History { get; set; }

	public void Main()
	{
	}
}
```

**Comment by adamabdelhamed** [comment]
I would recommend that you make your 'All' property a bool (like in your second block of code) and then after you have your parsed arguments do something like this: 

    if(myAppArgs.All)
    {
       IEnumerable<IFoo> = GetFoos();
    }
    else
    {
        IFoo specificFoo = GetFooById(myAppArgs.Id);
    }

You can be sure that the end user specified either the Id or the All flag since the conditional attributes you're using will enforce the conditions.

If you really want to separate this 'pseudo parsing' logic from whatever your program does then you can try to play around with an ArgHook, but it would probably end up being more complicated than just doing the if check yourself in your non static Main().

PS - It's cool to see people using the conditionally required and can't be combined with features.  Have you used it in multiple projects?  Any feedback? 

Thanks,
Adam

**Comment by Geraint87** [comment]
Thanks, I thought that would probably be the case. I think what I'll end up with is the `if-else`, as you suggested, but with just the one IEnumerable<IFoo> property, which is populated either way (a collection of Foos if All is specified, a Collection of one Foo if Id is specified).

This is the first time I've used the conditionally required features. In the past I've had similar needs, but I ended up splitting them them out into ActionMethods instead, as they were separate enough to warrant it.

All in all, I like them! I like the fluent-style readability and ease of use, and I think it suits my scenario really well. 
Downsides maybe, would be that the complexity increases really quickly, so if I had any more conditional arguments, or mutually exclusive sets, it gets very hard to follow!
And, rather than string property names, I would prefer it if the dependencies could be set up in code, with lambda `() => this.History` instead, for refactoring, or appearing in code lens `3 references` etc.

One thing I always find myself doing after parsing is re-wrapping the parsed output into another class. E.g. There could be 5 properties on my ApplicationArgs class, and some will be populated, some not, some used as a switch, or input to calculate other properties with an ArgIgnore on them...

What would be really nice is if you could have an input Args and a separate _parsed_ Args

e.g
```
public class ApplicationArgs
{
    [ArgShortcut("-i")]
    [ArgRequired(IfNot = "All")]
    [ArgCantBeCombinedWith("All")]
    public int Id { get; set; }

    [ArgShortcut("-a")]
    public bool All { get; set; }

    [ArgShortcut("d")]
    [ArgCantBeCombinedWith("History")]
    public DateTime? Date { get; set; }

    [ArgShortcut("-h")]
    [ArgCantBeCombinedWith("All")]
    public bool History { get; set; }
}
```
and then:
```
public class ApplicationParsedArgs
{
    IEnumerable<IFoo> Things { get; set; }
    
    DateTime StartDate { get; set; }
   
    DateTime EndDate { get; set; }
}
```
Where all my properties have been properly initialised now, and I have no optional fields.

Is this something that you think is possible? Or something that PowerArgs would be moving towards?

**Comment by adamabdelhamed** [comment]
Have you looked at the actions features described at the end of the readme's overview?  It lets you separate your program into distinct commands where each command can define its own inputs.  

I've found that if you factor your program this way then there is likely to be less conditional complexity within each command.  If you're finding that you have too many required if's or 'can't be combined withs' then you might want to split up your action into multiple commands.
